xml-format/doc/xml-format.1
.\" Automatically generated by Pandoc 2.3.1
.\"
.TH "xml\-format" "1" "2019\-03\-08" "" "xml\-utils"
.hy
.SH NAME
.PP
xml\-format \- Format XML files
.SH SYNOPSIS
.IP
.nf
\f[C]
$\ xml\-format\ [\-fOwh?]\ [\-i\ <str>]\ [\-o\ &lt;path&gt;]\ [<file>...]
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[I]xml\-format\f[] utility pretty prints an XML document by
removing ignorable blank nodes and indenting the XML tree.
.PP
A text node is considered ignorable if:
.IP \[bu] 2
its parent does not set or inherit xml:space as "preserve"
.IP \[bu] 2
its contents are entirely whitespace
.IP \[bu] 2
all its sibling text nodes are also ignorable
.IP \[bu] 2
all its ancestors contained only ignorable text nodes
.SH OPTIONS
.TP
.B \-f
Overwrite input XML files.
.RS
.RE
.TP
.B \-h \-?
Show usage message.
.RS
.RE
.TP
.B \-i <str>
Use <str> to indent each level of the XML tree.
The default is two spaces.
.RS
.RE
.TP
.B \-O
Omit the XML declaration from the formatted XML output.
.RS
.RE
.TP
.B \-o <path>
Output formatted XML to <path> instead of stdout.
.RS
.RE
.TP
.B \-w
Preserve elements which contain only whitespace.
Otherwise, these are converted to empty elements.
.RS
.RE
.TP
.B \-\-version
Show version information.
.RS
.RE
.TP
.B <file>...
XML file(s) to format.
If none are specified, the utility will read from stdin.
.RS
.RE
.SH EXAMPLE
.IP
.nf
\f[C]
$\ cat\ doc.xml
<section>\ \ \ \ \ <title>Example</title>

<para><emphasis>A</emphasis>\ <emphasis>B</emphasis>\ C</para>

<empty>\ \ \ \ </empty>

<p><b>A</b>\ <b>B</b>\ <b>C</b></p>
<p\ xml:space="preserve"><b>A</b>\ <b>B</b>\ <b>C</b></p>

<para>
\ \ \ \ See\ the\ following\ list:
\ \ \ \ <randomList>
\ \ \ \ \ \ <listItem><para>A</para></listItem>
\ \ \ \ \ \ <listItem><para>B</para></listItem>
\ \ \ \ \ \ <listItem><para>C</para></listItem>
\ \ \ \ </randomList>
</para>

</section>

$\ xml\-format\ test.xml
<section>
\ \ <title>Example</title>
\ \ <para><emphasis>A</emphasis>\ <emphasis>B</emphasis>\ C</para>
\ \ <empty/>
\ \ <p>
\ \ \ \ <b>A</b>
\ \ \ \ <b>B</b>
\ \ \ \ <b>C</b>
\ \ </p>
\ \ <p\ xml:space="preserve"><b>A</b>\ <b>B</b>\ <b>C</b></p>
\ \ <para>
\ \ \ \ See\ the\ following\ list:
\ \ \ \ <randomList>
\ \ \ \ \ \ <listItem><para>A</para></listItem>
\ \ \ \ \ \ <listItem><para>B</para></listItem>
\ \ \ \ \ \ <listItem><para>C</para></listItem>
\ \ \ \ </randomList>
\ \ </para>
</section>
\f[]
.fi
.SH AUTHORS
khzae.net.


xml-format/xml-format.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <libxml/tree.h>

#define PROG_NAME "xml-format"
#define VERSION "1.2.1"

/* Formatter options */
#define FORMAT_OVERWRITE	0x01
#define FORMAT_KEEPWSONLY	0x02

/* Determine if an option is set. */
bool optset(int opts, int opt)
{
	return ((opts & opt) == opt);
}

/* The blank text node children in an element are considered removable only if
 * ALL the text node children of that element are blank (no mixed content).
 *
 * If FORMAT_KEEPWSONLY is set, an element with a single blank text node child
 * will not be convereted to an empty element.
 */
bool blanks_are_removable(xmlNodePtr node, int opts)
{
	xmlNodePtr cur;
	int i;

	if (xmlNodeGetSpacePreserve(node) == 1) {
		return false;
	}

	for (cur = node->children, i = 0; cur; cur = cur->next, ++i) {
		if (cur->type == XML_TEXT_NODE && !xmlIsBlankNode(cur)) {
			return false;
		}
	}

	return i > 1 || !optset(opts, FORMAT_KEEPWSONLY);
}

/* Remove blank children. */
void remove_blanks(xmlNodePtr node)
{
	xmlNodePtr cur;

	cur = node->children;

	while (cur) {
		xmlNodePtr next;
		next = cur->next;

		if (xmlIsBlankNode(cur)) {
			xmlUnlinkNode(cur);
			xmlFreeNode(cur);
		}

		cur = next;
	}
}

/* Format XML nodes. */
void format(xmlNodePtr node, int opts)
{
	xmlNodePtr cur;

	if (blanks_are_removable(node, opts)) {
		remove_blanks(node);

		for (cur = node->children; cur; cur = cur->next) {
			format(cur, opts);
		}
	}
}

/* Format an XML file. */
void format_file(const char *path, const char *out, int opts)
{
	xmlDocPtr doc;

	if (!(doc = xmlReadFile(path, NULL, 0))) {
		return;
	}

	format(xmlDocGetRootElement(doc), opts);

	if (out) {
		xmlSaveFormatFile(out, doc, 1);
	} else if (optset(opts, FORMAT_OVERWRITE)) {
		xmlSaveFormatFile(path, doc, 1);
	} else {
		xmlSaveFormatFile("-", doc, 1);
	}

	xmlFreeDoc(doc);
}

/* Show usage message. */
void show_help(void)
{
	puts("Usage: " PROG_NAME " [-fwh?] [-i <str>] [-o <path>] [<file>...]");
	puts("");
	puts("Options:");
	puts("  -f         Overwrite input XML files.");
	puts("  -h -?      Show usage message.");
	puts("  -i <str>   Set the indentation string.");
	puts("  -o <path>  Output to <path> instead of stdout.");
	puts("  -w         Preserve elements containing only whitespace.");
	puts(" --version   Show version information.");
	puts("  <file>     XML file(s) to format. Otherwise, read from stdin.");
}

/* Show version information. */
void show_version(void)
{
	printf("%s (xml-utils) %s\n", PROG_NAME, VERSION);
	printf("Using libxml %s\n", xmlParserVersion);
}

int main(int argc, char **argv)
{
	int i;
	const char *sopts = "fi:o:wh?";
	struct option lopts[] = {
		{"version", no_argument, 0, 0},
		{0, 0, 0, 0}
	};
	int loptind = 0;

	int opts = 0;
	char *indent = NULL;
	char *out = NULL;

	while ((i = getopt_long(argc, argv, sopts, lopts, &loptind)) != -1) {
		switch (i) {
			case 0:
				if (strcmp(lopts[loptind].name, "version") == 0) {
					show_version();
					return 0;
				}
				break;
			case 'f':
				opts |= FORMAT_OVERWRITE;
				break;
			case 'i':
				indent = strdup(optarg);
				break;
			case 'o':
				out = strdup(optarg);
				break;
			case 'w':
				opts |= FORMAT_KEEPWSONLY;
				break;
			case 'h':
			case '?':
				show_help();
				return 0;
		}
	}

	if (indent) {
		xmlTreeIndentString = indent;
	}

	if (optind < argc) {
		for (i = optind; i < argc; ++i) {
			format_file(argv[i], out, opts);
		}
	} else {
		format_file("-", out, opts);
	}

	free(indent);
	free(out);

	return 0;
}
